File: docusign/templates.js
// docusign/templates.js
// ─────────────────────────────────────────────────────────────────────────────
// Wrappers around the DocuSign Templates API. Templates are reusable envelope
// definitions stored in your DocuSign account (documents + roles + tabs). You
// send one by supplying just the recipient who fills each role.
// ─────────────────────────────────────────────────────────────────────────────

import { apiFetch } from './client.js';

/** List the account's templates. */
export function listTemplates() {
  return apiFetch('/templates', { method: 'GET' });
}

/** Fetch one template's full definition. */
export function getTemplate(templateId) {
  return apiFetch(`/templates/${templateId}`, { method: 'GET' });
}

/**
 * Create + send an envelope from a stored template by binding roles to signers.
 * @param {object} p
 * @param {string} p.templateId
 * @param {string} p.emailSubject
 * @param {Array<{roleName, name, email, clientUserId?}>} p.roles
 */
export function sendFromTemplate({ templateId, emailSubject, roles }) {
  return apiFetch('/envelopes', {
    method: 'POST',
    body: {
      templateId,
      emailSubject,
      status: 'sent',
      templateRoles: roles.map((r) => ({
        roleName: r.roleName,
        name: r.name,
        email: r.email,
        ...(r.clientUserId ? { clientUserId: r.clientUserId } : {}),
      })),
    },
  });
}

File: backend/app/docusign/templates.py
"""Wrappers around the DocuSign Templates API (live calls)."""
from .client import api_fetch


async def list_templates() -> dict:
    return await api_fetch("/templates", {"method": "GET"})


async def get_template(template_id: str) -> dict:
    return await api_fetch(f"/templates/{template_id}", {"method": "GET"})


async def send_from_template(*, template_id: str, email_subject: str, roles: list) -> dict:
    return await api_fetch("/envelopes", {
        "method": "POST",
        "body": {
            "templateId": template_id,
            "emailSubject": email_subject,
            "status": "sent",
            "templateRoles": [{
                "roleName": r.get("roleName"), "name": r.get("name"), "email": r.get("email"),
                **({"clientUserId": r["clientUserId"]} if r.get("clientUserId") else {}),
            } for r in roles],
        },
    })
